www/src/pages/table/forms/TransformationForm.tsx

```diff
@@ -0,0 +1,289 @@
+                  Reset
+                </Button>
+              </Form.Item>
+            </Col>
+            <Col span={11} style={{ paddingLeft: "2rem" }}>
+              <Form.Item
+                name="onerror"
+                label="on error"
+                rules={[{ required: true }]}
+                initialValue={4}
+              >
+                <Radio.Group>
+                  <Radio value={1}>Set To Blank</Radio>
+                  <Radio value={2}>Store Error</Radio>
+                  <Radio value={3}>Keep Original</Radio>
+                  <Radio value={4}>Abort</Radio>
+                </Radio.Group>
+              </Form.Item>
+            </Col>
+            <Space></Space>
+            <Col span={4}>
+              <Form.Item
+                name="tolerance"
+                label="tolerance"
+                rules={[{ required: true }]}
+                initialValue={3}
+              >
+                <InputNumber placeholder="3" />
+              </Form.Item>
+            </Col>
+            <Col span={3}>
+              <Form.Item name="rows" label="rows">
+                <InputNumber />
+              </Form.Item>
+            </Col>
+          </Row>
+          <Row>
+            <Col span={24}>
+              {Array.isArray(result) ? (
+                <Table
+                  style={{ border: "1px solid #ccc" }}
+                  columns={columns}
+                  dataSource={result.map(transformationResult2Row)}
+                  pagination={{ pageSize: 4 }}
+                />
+              ) : result != undefined ? (
+                <Tag style={{ padding: 0 }} color={"volcano"}>
+                  <pre style={{ margin: 5 }}>{result}</pre>
+                </Tag>
+              ) : (
+                <></>
+              )}
+            </Col>
+          </Row>
+        </Form>
+      );
+    }
+  )
+);
+
+export const CustomEditor = ({
+  value,
+  onChange,
+}: {
+  value?: string;
+  onChange?: (value: string | undefined) => void;
+}) => {
+  const options = {
+    autoIndent: "full",
+    contextmenu: true,
+    fontFamily: "monospace",
+    fontSize: 13,
+    lineHeight: 24,
+    matchBrackets: "always",
+    minimap: {
+      enabled: false,
+    },
+    scrollbar: {
+      horizontalSliderSize: 4,
+      verticalSliderSize: 18,
+    },
+    selectOnLineNumbers: true,
+    roundedSelection: false,
+    readOnly: false,
+    cursorStyle: "line",
+    automaticLayout: true,
+  };
+
+  return (
+    <div style={{ border: "1px solid #ccc" }}>
+      <Editor
+        height="25vh"
+        defaultLanguage="python"
+        onChange={onChange}
+        options={options}
+        value={value}
+      />
+    </div>
+  );
+};
+
+function transformationResult2Row(tbl: TransformationResult) {
+  return {
+    key: tbl.path,
+    row_id: tbl.path + 1,
+    previous_value: tbl.value,
+    result: tbl.ok !== undefined ? tbl.ok.toString() : tbl.error.trim(),
```

Comment:
table.ok is annotated as string, why we need to .toString? and if it can be undefined, we should update its type as well.

Reply:
we return the values of true/false as boolean values to handle those cases I have added this.

Reply:
so you need to update the type.

Reply:
updated the type